integration/ct_integration.go

```diff
@@ -0,0 +1,636 @@
+	// Calculate leaf hash =  SHA256(0x00 | tls-encode(MerkleTreeLeaf))
+	leaf = ct.MerkleTreeLeaf{
+		Version:  ct.V1,
+		LeafType: ct.TimestampedEntryLeafType,
+		TimestampedEntry: &ct.TimestampedEntry{
+			Timestamp: precertSCT.Timestamp,
+			EntryType: ct.PrecertLogEntryType,
+			PrecertEntry: &ct.PreCert{
+				IssuerKeyHash:  sha256.Sum256(issuer.RawSubjectPublicKeyInfo),
+				TBSCertificate: tbs,
+			},
+			Extensions: precertSCT.Extensions,
+		},
+	}
+	leafData, err := tls.Marshal(leaf)
+	if err != nil {
+		return fmt.Errorf("tls.Marshal(precertLeaf)=(nil,%v); want (_,nil)", err)
+	}
+	hash := sha256.Sum256(append([]byte{merkletree.LeafPrefix}, leafData...))
+	rsp, err := pool.Pick().GetProofByHash(ctx, hash[:], sthN1.TreeSize)
+	stats.done(ctfe.GetProofByHashName, 200)
+	if err != nil {
+		return fmt.Errorf("got GetProofByHash(precertSCT, size=%d)=nil,%v", sthN1.TreeSize, err)
+	}
+	if rsp.LeafIndex != precertIndex {
+		return fmt.Errorf("got GetProofByHash(precertSCT, size=%d).LeafIndex=%d; want %d", sthN1.TreeSize, rsp.LeafIndex, precertIndex)
+	}
+	fmt.Printf("%s: Inclusion proof leaf %d @ %d -> root %d = %x\n", cfg.Prefix, precertIndex, precertSCT.Timestamp, sthN1.TreeSize, rsp.AuditPath)
+	if err := Verifier.VerifyInclusionProof(rsp.LeafIndex, int64(sthN1.TreeSize), rsp.AuditPath, sthN1.SHA256RootHash[:], leafData); err != nil {
+		return fmt.Errorf("got VerifyInclusionProof(%d,%d,...)=%v; want nil", precertIndex, sthN1.TreeSize, err)
+	}
+	if err := stats.check(cfg, servers); err != nil {
+		return fmt.Errorf("unexpected stats check: %v", err)
+	}
+
+	// Stage 15: invalid consistency proof
+	if rsp, err := pool.Pick().GetSTHConsistency(ctx, 2, 299); err == nil {
+		return fmt.Errorf("got GetSTHConsistency(2,299)=(%+v,nil); want (nil,_)", rsp)
+	}
+	stats.done(ctfe.GetSTHConsistencyName, 400)
+	fmt.Printf("%s: GetSTHConsistency(2,299)=(nil,_)\n", cfg.Prefix)
+
+	// Stage 16: invalid inclusion proof
+	wrong := sha256.Sum256([]byte("simply wrong"))
+	if rsp, err := pool.Pick().GetProofByHash(ctx, wrong[:], sthN1.TreeSize); err == nil {
+		return fmt.Errorf("got GetProofByHash(wrong, size=%d)=(%v,nil); want (nil,_)", sthN1.TreeSize, rsp)
+	}
+	stats.done(ctfe.GetProofByHashName, 400)
+	fmt.Printf("%s: GetProofByHash(wrong,%d)=(nil,_)\n", cfg.Prefix, sthN1.TreeSize)
+
+	return nil
+}
+
+// TimeFromMS converts a timestamp in milliseconds (as used in CT) to a time.Time.
+func TimeFromMS(ts uint64) time.Time {
+	secs := int64(ts / 1000)
+	msecs := int64(ts % 1000)
+	return time.Unix(secs, msecs*1000000)
+}
+
+// SignatureToString formats a CT signature for display.
+func SignatureToString(signed *ct.DigitallySigned) string {
+	return fmt.Sprintf("Signature: Hash=%v Sign=%v Value=%x", signed.Algorithm.Hash, signed.Algorithm.Signature, signed.Signature)
+}
+
+// GetChain retrieves a certificate from a file of the given name and directory.
+func GetChain(dir, path string) ([]ct.ASN1Cert, error) {
```

Comment:
There are a lot of exported helper funcs in here, do they need to be exported?
I'm thinking that _if so_ then they could perhaps live somewhere else.